rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -1274,104 +1349,76 @@ RMW_Connext_Subscriber::create(
     RMW_CONNEXT_LOG_ERROR_SET("failed to create DDS reader")
     return nullptr;
   }
-
-  auto scope_exit_dds_reader_delete =
-    rcpputils::make_scope_exit(
-    [sub, dds_reader]()
-    {
-      if (DDS_RETCODE_OK !=
-      DDS_Subscriber_delete_datareader(sub, dds_reader))
-      {
-        RMW_CONNEXT_LOG_ERROR_SET(
-          "failed to delete DDS DataWriter")
-      }
-    });
-
-  RMW_Connext_Subscriber * rmw_sub_impl =
-    new (std::nothrow) RMW_Connext_Subscriber(
-    ctx,
-    dds_reader,
-    topic,
-    type_support,
-    subscriber_options->ignore_local_publications,
-    topic_created,
-    cft_topic,
-    internal);
-
-  if (nullptr == rmw_sub_impl) {
-    RMW_CONNEXT_LOG_ERROR_SET("failed to allocate RMW subscriber")
-    return nullptr;
-  }
-  scope_exit_type_unregister.cancel();
-  scope_exit_topic_delete.cancel();
-  scope_exit_dds_reader_delete.cancel();
-
-  return rmw_sub_impl;
+  return dds_reader;
 }
 
 rmw_ret_t
-RMW_Connext_Subscriber::finalize()
+RMW_Connext_Subscriber::finalize(const bool reset_cft)
 {
   RMW_CONNEXT_LOG_DEBUG_A(
     "finalizing subscriber: sub=%p, type=%s",
     (void *)this, this->type_support->type_name())
 
   // Make sure subscriber's condition is detached from any waitset
-  this->status_condition.invalidate();
+  this->status_condition->invalidate();
 
-  if (this->loan_len > 0) {
-    this->loan_next = this->loan_len;
-    if (RMW_RET_OK != this->return_messages()) {
-      return RMW_RET_ERROR;
+  {
+    std::lock_guard<std::mutex> lock(this->loan_mutex);
+    if (this->loan_len > 0) {
+      if (RMW_RET_OK != this->return_messages()) {
+        return RMW_RET_ERROR;
+      }
     }
   }
 
-  if (DDS_RETCODE_OK !=
+  if (nullptr != this->dds_reader && DDS_RETCODE_OK !=
     DDS_Subscriber_delete_datareader(
       this->dds_subscriber(), this->dds_reader))
   {
     RMW_CONNEXT_LOG_ERROR_SET("failed to delete DDS DataReader")
     return RMW_RET_ERROR;
   }
```

Comment:
This block fails to reset `this->dds_reader`, this is a problem if we assume `finalize()` is idempotent.

I suggest refactoring to:

```cpp
if (nullptr != this->dds_reader) {
  if (DDS_RETCODE_OK !=
    DDS_Subscriber_delete_datareader(this->dds_subscriber(), this->dds_reader))
  {
    RMW_CONNEXT_LOG_ERROR_SET("failed to delete DDS DataReader")
    return RMW_RET_ERROR;
  }
  this->dds_reader = nullptr;
}
```

You could also just always reset `this->dds_reader`, but I think the nested block is more readable and more consistent with the rest of the the code.

Reply:
Putting `this->dds_reader = nullptr;` after `DDS_Subscriber_delete_datareader` seems reasonable.